src/inter_dc/reqrouter/inter_dc_query_req.erl

```diff
@@ -0,0 +1,248 @@
+
+            {reply, ok, State#state{req_id = ReqId + 1}};
+        _ ->
+            ?LOG_WARNING("Could not find ~p:~p in socket dict ~p", [DCID, Partition, State#state.sockets]),
+            {reply, unknown_dc, State}
+    end.
+
+handle_info({'EXIT', Pid, Reason}, State) ->
+    ?LOG_NOTICE("Connect query socket ~p shutdown: ~p", [Pid, Reason]),
+    {noreply, State};
+handle_info(_, State) ->
+    {noreply, State}.
+
+terminate(_Reason, State) ->
+    F = fun({_, Socket}) -> inter_dc_utils:close_socket(Socket) end,
+    lists:foreach(F, dict:to_list(State#state.sockets)),
+    ok.
+
+handle_cast(_Request, State) -> {noreply, State}.
+
+code_change(_OldVsn, State, _Extra) -> {ok, State}.
+
+
+%%%% Internal methods ---------------------------------------------------------+
+
+-spec del_dc(dcid(), req_dict()) -> req_dict().
+del_dc(DCID, Dict) ->
+    %% filter all DCID-Partition pairs to remove them
+    MatchingDCID = fun({DictDCID, _DictPartition}, _Value) -> DictDCID == DCID end,
+    ToRemoveDict = dict:filter(MatchingDCID, Dict),
+
+    %% close sockets and erase entry of input dict
+    RemoveDCIDPartitionEntry =
+        fun({Key, Socket}, AccDict) ->
+            %% the sockets are the same, but close all of them anyway (close socket is idempotent)
+            inter_dc_utils:close_socket(Socket),
+            dict:erase(Key, AccDict)
+        end,
+
+    lists:foldl(RemoveDCIDPartitionEntry, Dict, dict:to_list(ToRemoveDict)).
+
+%% A node is a list of addresses because it can have multiple interfaces
+%% this just goes through the list and connects to the first interface that works
+-spec connect_to_node([socket_address()]) -> {ok, zmq_socket()} | connection_error.
+connect_to_node([]) ->
+    ?LOG_ERROR("Unable to subscribe to DC log reader"),
+    connection_error;
+connect_to_node([_Address = {Ip, Port}| Rest]) ->
+    %% Test the connection
+    case connect_req(Ip, Port) of
+        {ok, Socket1} ->
+            %% Always use 0 as the id of the check up message
+            ReqIdBinary = inter_dc_txn:req_id_to_bin(0),
+            BinaryVersion = ?MESSAGE_VERSION,
+
+            ok = chumak:send(Socket1, <<BinaryVersion/binary, ReqIdBinary/binary, ?CHECK_UP_MSG>>),
+            Response = chumak:recv(Socket1),
+            inter_dc_utils:close_socket(Socket1),
+
+            case Response of
+                {ok, Binary} ->
+                    %% check that an ok msg was received
+                    {_, <<?OK_MSG>>} = inter_dc_utils:check_version_and_req_id(Binary),
+                    %% Create a subscriber socket for the specified DC
+                    case connect_req(Ip, Port) of
+                        {ok, Socket} ->
+                            {ok, Socket};
+                        _ ->
+                            connect_to_node(Rest)
+                    end;
+                _ ->
+                    connect_to_node(Rest)
+            end;
+        _ ->
+            connect_to_node(Rest)
+    end.
```

Comment:
Question: Do we have to close the socket and open it again? Couldn't we just return the socket we opened initially and depending on the response close it?